### ISSUE TITLE
BUG: concat() with datetime variable produces unclear/misleading error

### ISSUE THREAD

**Issue by MichalRIcar** [pandas-dev/pandas#55132]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
from pandas import DataFrame as pDF
from datetime import datetime

pd.concat([
    pDF({'A': datetime.strptime('23/09/14', "%y/%m/%d")}, index=[0]),
    pDF({'A': datetime.strptime('23/09/14', "%y/%m/%d")}, index=[0]).astype('datetime64[ns]')
])
```

### Issue Description

The example above raises a 'ValueError: Shape of passed values is (1, 1), indices imply (2, 1).' 

However, this error occurs because the datetime variables have different precisions; one is datetime64[us], and the other is datetime64[ns].  

I assume that the concatenation operation intends to raise an error in this case, even though similar cases with concatenating float and int values do not raise an error and join datasets by casting precision from int to float. Therefore, I am considering whether to handle datetime concatenation in the same way and adjust it accordingly..?

### Expected Behavior

Either don't raise and concat while casting precision, similarly as int goes to float, or raise an error but with proper reasoning.

### Installed Versions

<details>
python              : 3.11.4.final.0
pandas              : 2.1.0

</details>

**Comment by paulreece** [comment]
I can confirm that the Error is raised on the above test case.

**Comment by jbrockmendel** [comment]
closed by #53641?

**Comment by Charlie-XIAO** [comment]
@jbrockmendel I can confirm that this bug no longer exists on the main branch, due to #53641.

**Comment by lithomas1** [comment]
This should've been fixed in 2.2 I think.
